private-identity-audit/src/prover.rs

```diff
@@ -0,0 +1,183 @@
+//! Given C = g^x * h^y * f^z
+//!
+//! The goal is to prove the following statements.
+//! - a = C^r -> ZKP(a; C)
+//! - b = h^{y*r} * f^{z*r} -> ZKP(b; h, f)
+//! - a/b = g^{x*r} -> ZKP(a/b; g)
+
+use crate::{
+    errors::Fallible,
+    proofs::{apply_challenge, generate_initial_message},
+    ChallengeResponder, CommittedUids, ProofGenerator, Proofs, ProverFinalResponse, ProverSecrets,
+};
+use confidential_identity::{pedersen_commitments::PedersenGenerators, CddClaimData};
+use cryptography_core::curve25519_dalek::{ristretto::RistrettoPoint, scalar::Scalar};
+use rand::seq::SliceRandom;
+use rand_core::{CryptoRng, RngCore};
+use sha3::{Digest, Sha3_512};
+
+pub struct InitialProver;
+pub struct FinalProver;
+
+/// Modified version of `generate_pedersen_commit` function of Confidential Identity Library.
+pub fn generate_blinding_factor(uid: Scalar, did: Scalar) -> Scalar {
+    let hash = Sha3_512::default()
+        .chain(uid.as_bytes())
+        .chain(did.as_bytes());
+    Scalar::from_hash(hash)
+}
+
+impl ProofGenerator for InitialProver {
+    fn generate_initial_proofs<T: RngCore + CryptoRng>(
+        claim: CddClaimData,
+        rng: &mut T,
+    ) -> Fallible<(ProverSecrets, Proofs)> {
+        let pg = PedersenGenerators::default();
+        let blinding_factor =
+            generate_blinding_factor(claim.investor_unique_id, claim.investor_did);
+        let secrets = [
+            claim.investor_unique_id,
+            claim.investor_did,
+            blinding_factor,
+        ];
+        let cdd_id = pg.commit(&secrets);
+
+        let r = Scalar::random(rng);
```

Comment:
This `r` here is technically the challenge. Is the protocol secure if we rely on the prover to generate a random challenge (they could technically use a scalar they choose), or should we use a Merlin transcript here?